### PR TITLE
[JIT] Implement staged symbolics for pack_padded/pad_packed

### PR DESCRIPTION
**Pull request by jamesr66a**
Previously, the symbolics for these functions were specialized to ONNX only. This PR makes it so that they instead emit a PythonOp with an embedded reference to the ONNX symbolic. This allows the interpreter to run these as Python, while ONNX export still retains the information it needs to emit its placeholder operators

### CODE REVIEW

File: torch/jit/__init__.py [hunk 1]
```diff
@@ -804,3 +806,165 @@ def forward(self, *args, **kwargs):
 
 if not torch._C._jit_init():
     raise RuntimeError("JIT initialization failed")
+
+
+def _scalar(x):
+    """Convert a scalar tensor into a Python value."""
```
Comment (apaszke) [inline torch/jit/__init__.py]: This is pretty confusing, because we usually say "scalar tensor" when we mean a 0d tensor

File: torch/jit/__init__.py [hunk 2]
```diff
@@ -804,3 +806,165 @@ def forward(self, *args, **kwargs):
...
+    n = g.create(ns_opname, args, outputs)
+    for k, v in sorted(kwargs.items()):
+        # TODO: enable inplace in aten exporting mode.
+        if k == "inplace":
+            continue
+        _add_attribute(n, k, v, aten=aten)
+    return n
+
+
+def _graph_op(g, opname, *raw_args, **kwargs):
```
Comment (apaszke) [inline torch/jit/__init__.py]: 1. This probably shouldn't be called `_graph_op`, because there's no reason to not call ATen operators "ops". `_graph_onnx` would be more consistent with `_graph_at`

File: torch/jit/__init__.py [hunk 3]
```diff
@@ -804,3 +806,165 @@ def forward(self, *args, **kwargs):
...
     raise RuntimeError("JIT initialization failed")
+
+
+def _scalar(x):
+    """Convert a scalar tensor into a Python value."""
+    assert x.numel() == 1
+    return x[0]
+
+
+def _is_onnx_list(value):
```
Comment (apaszke) [inline torch/jit/__init__.py]: Why did those functions migrate from `torch/onnx` to `torch/jit`? What was wrong? I'd rather avoid keeping things like `_is_onnx_list` here.

File: torch/jit/__init__.py [hunk 4]
```diff
@@ -804,3 +806,165 @@ def forward(self, *args, **kwargs):
...
+
+    NB: This is monkey-patched onto Node.
+    """
+    sel = self.kindOf(k)
+    return getattr(self, sel)(k)
+
+
+torch._C.Graph.op = _graph_op
+torch._C.Graph.at = _graph_at
+torch._C.Graph.constant = _graph_constant
```
Comment (apaszke) [inline torch/jit/__init__.py]: Having `.constant` function that creates an ONNX node, inside the JIT directory (where we usually deal with ATen and builtin JIT ops, which also have a `Constant` node is very confusing and error prone.

File: torch/jit/__init__.py [hunk 5]
```diff
@@ -804,3 +806,165 @@ def forward(self, *args, **kwargs):
...
+    if m is None:
+        raise IndexError((
+            "Invalid attribute specifier '{}' names " +
+            " must be suffixed with type, e.g. 'dim_i' or 'dims_i'").format(key))
+    name, kind = m.group(1), m.group(2)
+    if _is_onnx_list(value):
+        kind += "s"
+    if aten:
+        if torch.is_tensor(value):
+            # Caffe2 proto does not support tensor attribute.
```
Comment (apaszke) [inline torch/jit/__init__.py]: Why do we care about Caffe2 proto here?

File: torch/jit/__init__.py [hunk 6]
```diff
@@ -804,3 +806,165 @@ def forward(self, *args, **kwargs):
...
+            if value.numel() > 1:
+                raise ValueError("Should not pass tensor attribute")
+            value = _scalar(value)
+            if isinstance(value, float):
+                kind = "f"
+            else:
+                kind = "i"
```
Comment (apaszke) [inline torch/jit/__init__.py]: Having this auto-inference is also confusing when you take into account that we usually use the kwarg suffix to determine the type. I'd rather avoid having this in the JIT.

File: torch/csrc/jit/python_ir.cpp [hunk 7]
```diff
@@ -21,6 +22,22 @@ void initPythonIRBindings(PyObject * module_) {
...
+      std::string cconv(inputs.size(), 't');
+      func.attr("symbolic") = symbolic;
+      Node* new_node = g.insertNode(g.createPythonOp(
+        THPObjectPtr(func.release().ptr()), cconv, false, {}, {}, false));
+      for (auto i : inputs)
+        new_node->addInput(i);
+      std::vector<Value*> outputs;
+      for (size_t i = 0; i < n_outputs; ++i)
+        new_node->addOutput();
+#ifndef NO_PYTHON
```
Comment (apaszke) [inline torch/csrc/jit/python_ir.cpp]: I doubt you'll have `NO_PYTHON` defined inside a `pybind` function 😉

File: torch/csrc/jit/python_ir.cpp [hunk 8]
```diff
@@ -21,6 +22,22 @@ void initPythonIRBindings(PyObject * module_) {
       ss << g;
       return ss.str();
     })
+    .def("wrapPyFuncWithSymbolic", [](Graph &g, py::function func, std::vector<Value*> inputs, size_t n_outputs, py::function symbolic) {
```
Comment (apaszke) [inline torch/csrc/jit/python_ir.cpp]: Docs for this function? What's the expected usage and what does it do?

File: torch/jit/__init__.py (hunk 3 above)
Comment (jamesr66a) [reply torch/jit/__init__.py]: This was required during my early investigation since the monkey-patched .op() method was not present on `Graph`. Let me see if i can revert this move now that I have a (slightly different from my original) solution

File: torch/jit/__init__.py (hunk 5 above)
Comment (jamesr66a) [reply torch/jit/__init__.py]: Great question

**Comment by ezyang** [comment]
Is the non-ONNX trace supposed to run in the interpreter?

**Comment by jamesr66a** [comment]
@ezyang yeah

File: torch/nn/utils/rnn.py [hunk 9]
```diff
@@ -199,6 +202,27 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0):
...
+                                   batch_first=batch_first, padding_value=padding_value)
+
+    data, lengths = g.wrapPyFuncWithSymbolic(
+        pad_packed_sequence_trace_wrapper, [input.data, input.batch_sizes], 2,
+        _onnx_symbolic_pad_packed_sequence)
+    return data, lengths
+
+
+pad_packed_sequence = torch.onnx.symbolic_override_packed_sequence_based(
+    _symbolic_pad_packed_sequence)(pad_packed_sequence)
```
Comment (ezyang) [inline torch/nn/utils/rnn.py]: Man, I can't believe this passes lint lol.

File: torch/nn/utils/rnn.py [hunk 10]
```diff
@@ -199,6 +202,27 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0):
     return output, Variable(torch.LongTensor(lengths))
 
 
+def _symbolic_pad_packed_sequence(g, input, batch_first=False, padding_value=0.0):
```
Comment (ezyang) [inline torch/nn/utils/rnn.py]: This seems like a weirdly large amount of boilerplate compared to what we had before. What's going on?

**Comment by ezyang** [comment]
So... is the reason you have to hand-construct the PythonOp at tracing time, because if you actually made a PythonOp class, autograd would stop working? But isn't that still true if you try to run the traced thing?

**Comment by ezyang** [comment]
(I understand where you are headed with this patch but I am not sure how you actually make it work.)

**Comment by ezyang** [comment]
Offline discussion concluded that it should Just Work(TM), but it needs a test

File: torch/nn/utils/rnn.py (hunk 10 above)
Comment (jamesr66a) [reply torch/nn/utils/rnn.py]: So we need two wrappers:

1) A Python function that is called when we execute the PythonOp. This is a nested `def` which wraps the original operation so we can capture the python args `batch_first` and `padding_value`. Here this function is `pad_packed_sequence_trace_wrapper`
2) A symbolic function for the ONNX pass. This is yet another nested def here: `pad_packed_sequence_trace_wrapper`

**Review by zdevito — APPROVED** [review]
This looks fine. Keep in mind this is a work around to get rnn.py running in the hybrid frontend. In the future we want these overrides to provide graph-based implementations.

**Comment by zdevito** [comment]
@jamesr66a it is important to make sure onnx tests pass for this one, since it modifies that pathway.

File: test/test_jit.py [hunk 11]
```diff
@@ -2213,6 +2213,42 @@ def forward(self, v):
...
+                return x
+
+        x = torch.ones(T, B, C, requires_grad=True)
+        seq_lens = torch.ones(B, dtype=torch.int32)
+        m = PadPackedWrapper()
+        m_traced = torch.jit.trace(x, seq_lens)(m)
+        criterion = nn.MSELoss()
+
+        y = m(x, seq_lens)
+        loss = criterion(y, torch.zeros_like(y))
```
Comment (apaszke) [inline test/test_jit.py]: You can probably just use `.sum()` as a fake loss

File: torch/nn/utils/rnn.py [hunk 12]
```diff
@@ -137,17 +127,37 @@ def pack_padded_sequence(input, lengths, batch_first=False):
     return PackedSequence(data, batch_sizes)
 
 
-def _symbolic_pad_packed_sequence(g, input, batch_first=False, padding_value=0.0, total_length=None):
+def _onnx_symbolic_pack_padded_sequence(g, input, lengths):
+    return g.op("prim::PackPadded", input, lengths, outputs=2)
```
Comment (apaszke) [inline torch/nn/utils/rnn.py]: Is this used?

File: torch/nn/utils/rnn.py [hunk 13]
```diff
@@ -137,17 +127,37 @@ def pack_padded_sequence(input, lengths, batch_first=False):
...
-        data = g.op('Transpose', data, perm_i=[1, 0, 2])
-    return data, lengths
+        input = g.op('Transpose', input, perm_i=[1, 0, 2])
+    # There currently is no PackPadded operator in ONNX. We rely on an
+    # optimization pass to remove this later. It is an error if all
+    # PackPadded operators cannot be optimized out.
+
+    def _onnx_symbolic_pack_padded_sequence(g, input, lengths):
+        if batch_first:
+            input = g.op('Transpose', input, perm_i=[1, 0, 2])
```
Comment (apaszke) [inline torch/nn/utils/rnn.py]: Won't this duplicate transposes? You have one outside of this function already

**Review by apaszke — COMMENTED** [review]
LGTM. Two minor things

File: torch/nn/utils/rnn.py (hunk 12 above)
Comment (jamesr66a) [reply torch/nn/utils/rnn.py]: merge error

**Comment by jamesr66a** [comment]
@onnxbot retest this please

File: test/test_jit.py [hunk 14]
```diff
@@ -2213,6 +2213,41 @@ def forward(self, v):
...
+            def __init__(self):
+                super(PadPackedWrapper, self).__init__()
+
+            def forward(self, x, seq_lens):
+                x = pack_padded_sequence(x, seq_lens)
+                x, _ = pad_packed_sequence(x)
+                return x
+
+        x = torch.ones(T, B, C, requires_grad=True)
+        seq_lens = torch.ones(B, dtype=torch.int32)
```
Comment (apaszke) [inline test/test_jit.py]: Can we test on some other tensor? This one is extremely simple, and it would also be good to assert that the output is equal to input (since those functions are meant to be inverses of each other).

**Review by apaszke — APPROVED** [review]
